tests/e2e/tidbcluster/tidbcluster.go

```diff
@@ -327,6 +327,23 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 		framework.ExpectNoError(err, "failed to wait for TidbCluster ready: %q", tc.Name)
 	})
 
+	ginkgo.It("should direct upgrade tc successfully when PD replicas less than 2.", func() {
+		clusterName := "upgrade-cluster-pd-1"
+		tc := fixture.GetTidbCluster(ns, clusterName, utilimage.TiDBLatestPrev)
+		tc.Spec.PD.Replicas = 1
+		// Deploy
+		utiltc.MustCreateTCWithComponentsReady(genericCli, oa, tc, 10*time.Minute, 5*time.Second)
+		ginkgo.By(fmt.Sprintf("Upgrading tidb cluster from %s to %s", tc.Spec.Version, utilimage.TiDBLatest))
+		err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+			tc.Spec.Version = utilimage.TiDBLatest
+			return nil
+		})
+		framework.ExpectNoError(err, "failed to upgrade TidbCluster: %q", tc.Name)
+		err = oa.WaitForTidbClusterReady(tc, 10*time.Minute, 5*time.Second)
```

Comment:
still hard to check no `gracefulUpgrade` occurred?

Reply:
Yes, it's hard to check that, at least, the upgrade succeeds without adding annotation.